cosmic-panel-bin/src/notifications.rs

```diff
@@ -0,0 +1,179 @@
+    os::{
+        fd::{AsRawFd, FromRawFd, OwnedFd, RawFd},
+        unix::net::UnixStream,
+    },
+    sync::Arc,
+};
+
+use crate::{process::mark_as_cloexec, space_container::SpaceContainer};
+use anyhow::{Context, Result};
+use cosmic_notifications_util::{PanelRequest, PANEL_NOTIFICATIONS_FD};
+use sendfd::{RecvWithFd, SendWithFd};
+use smithay::reexports::{
+    calloop::{
+        self, channel::SyncSender, generic::Generic, Interest, LoopHandle, Mode, PostAction,
+    },
+    nix::{fcntl, unistd},
+};
+use tracing::{error, info, trace, warn};
+use xdg_shell_wrapper::shared_state::GlobalState;
+
+#[derive(Debug)]
+pub enum PendingAppletEvent {
+    Add(String, UnixStream),
+    Remove(String),
+}
+
+pub fn init(
+    loop_handle: &LoopHandle<GlobalState<SpaceContainer>>,
+) -> Result<SyncSender<PendingAppletEvent>> {
+    let fd_num = std::env::var(PANEL_NOTIFICATIONS_FD)?;
+    let fd = fd_num.parse::<RawFd>()?;
+    // set CLOEXEC
+    let flags = fcntl::fcntl(fd, fcntl::FcntlArg::F_GETFD);
+    let result = flags
+        .map(|f| fcntl::FdFlag::from_bits(f).unwrap() | fcntl::FdFlag::FD_CLOEXEC)
+        .and_then(|f| fcntl::fcntl(fd, fcntl::FcntlArg::F_SETFD(f)));
+    let mut daemon_socket = match result {
+        // CLOEXEC worked and we can startup with session IPC
+        Ok(_) => unsafe { UnixStream::from_raw_fd(fd) },
+        // CLOEXEC didn't work, something is wrong with the fd, just close it
+        Err(err) => {
+            let _ = unistd::close(fd);
+            return Err(err).with_context(|| "Failed to setup session socket");
+        }
+    };
+
+    daemon_socket
+        .set_nonblocking(true)
+        .expect("Couldn't set nonblocking");
+    // read remaining bytes from socket
+    {
+        let mut buf = [0u8; 128];
+        while let Ok(size) = daemon_socket.read(&mut buf) {
+            if size == 0 {
+                break;
+            }
+        }
+    }
+    let data = ron::ser::to_string(&PanelRequest::Init)?;
+    daemon_socket.write_all(format!("{}\n", data).as_bytes())?;
+    let daemon_socket: Arc<UnixStream> = Arc::new(daemon_socket);
+    let daemon_socket_clone = daemon_socket.clone();
+    info!("Inserting channel into event loop");
+
+    // insert channel for spaces to send pending nortification appplet to
+    let (tx, rx) = calloop::channel::sync_channel(10);
+
+    loop_handle
+        .insert_source(
+            rx,
+            move |msg: calloop::channel::Event<PendingAppletEvent>, _, state| match msg {
+                calloop::channel::Event::Msg(msg) => match msg {
+                    PendingAppletEvent::Add(id, stream) => {
+                        info!("Received pending notification applet for space {}", &id);
+                        if !state.space.notification_applet_spaces.contains(&id) {
+                            // state.space.pending_notification_applet_ids.push(msg);
+                            if let Err(err) = write_socket(&daemon_socket, state, (id, stream)) {
```

Comment:
Since the fd is set as non-blocking, this could return `EWOULDBLOCK`, right? Generally that should always be handled for non-blocking writes, though I guess that only happens if the socket's buffer is full, and I'm not sure how likely that is.